scrapy/http/request/form.py

```diff
@@ -33,7 +33,11 @@ def __init__(self, *args, **kwargs):
                 self.headers.setdefault(b'Content-Type', b'application/x-www-form-urlencoded')
                 self._set_body(querystr)
             else:
-                self._set_url(self.url + ('&' if '?' in self.url else '?') + querystr)
+                url_split = urlsplit(self.url)
+                formdata_key_list = list(dict(parse_qsl(querystr)).keys())
```

Comment:
Since this is used only for checking filtering in the next line, it would be more efficient with a set instead of a list:
```
formdata_keys = set(dict(parse_qsl(querystr)).keys())
```